sslr-core/src/main/java/com/sonar/sslr/api/typed/Input.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * SonarSource Language Recognizer
+ * Copyright (C) 2010 SonarSource
+ * dev@sonar.codehaus.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package com.sonar.sslr.api.typed;
+
+import com.google.common.annotations.Beta;
+import com.google.common.collect.Lists;
+
+import java.io.File;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.List;
+
+@Beta
+public class Input {
+
+  private static final URI FAKE_URI = new File("tests://unittests").toURI();
+
+  private final char[] input;
+  private final URI uri;
+  private final int[] newLineIndexes;
+
+  public Input(char[] input) {
+    this(input, FAKE_URI);
+  }
+
+  public Input(char[] input, URI uri) {
+    this.input = input;
+    this.uri = uri;
+
+    List<Integer> newLineIndexesBuilder = Lists.newArrayList();
+    for (int i = 0; i < input.length; i++) {
+      if (isNewLine(input, i)) {
+        newLineIndexesBuilder.add(i + 1);
+      }
+    }
+    this.newLineIndexes = new int[newLineIndexesBuilder.size()];
+    for (int i = 0; i < newLineIndexes.length; i++) {
+      this.newLineIndexes[i] = newLineIndexesBuilder.get(i);
+    }
+  }
+
+  public char[] input() {
+    return input;
+  }
+
+  public URI uri() {
+    return uri;
+  }
+
+  public String substring(int from, int to) {
+    StringBuilder sb = new StringBuilder();
+    for (int i = from; i < to; i++) {
+      sb.append(input[i]);
+    }
+    return sb.toString();
+  }
+
+  public int[] lineAndColumnAt(int index) {
+    int[] result = new int[2];
+    result[0] = lineAt(index);
+    result[1] = index - lineStartIndex(result[0]) + 1;
+    return result;
+  }
+
+  private int lineAt(int index) {
+    int i = Arrays.binarySearch(newLineIndexes, index);
+    return i >= 0 ? i + 2 : -i;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS864)